mmv1/templates/terraform/pre_update/alloydb_cluster.go.erb

```diff
@@ -0,0 +1,53 @@
+// Restrict modification of cluster_type from PRIMARY to SECONDARY as it is an invalid operation
+if d.HasChange("cluster_type") && d.Get("cluster_type") == "SECONDARY"{
+   return fmt.Errorf("Can not convert a primary cluster to a secondary cluster.")
+}
+
+// Implementation for cluster promotion
+if d.HasChange("cluster_type") && d.Get("cluster_type") == "PRIMARY"{
+
+    if !d.HasChange("secondary_config") || !tpgresource.IsEmptyValue(reflect.ValueOf(d.Get("secondary_config"))) {
+        return fmt.Errorf("Remove the secondary_config field to promote the cluster to primary cluster.")
+    }
+
+    if len(updateMask) > 2{
+        return fmt.Errorf("The only change when promoting a cluster should be \n1. Set cluster_type = PRIMARY \n2. Remove the secondary_config field from the config.")
```

Comment:
Terraform is generally expected to go from any valid configuration to any other valid configuration- this indicates a bug in the resource.